[2/4] Intel GPU Runtime Upstreaming for Device

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117619
* #117611
* #116869
* #116850
* __->__ #116833

# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), as mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

EikanWang commented on aten/src/ATen/xpu/detail/XPUHooks.cpp:
```diff
@@ -0,0 +1,35 @@
+  return "XPU backend";
+}
+
+int XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
```
```suggestion
DeviceIndex XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
```

guangyey replied:
`device_id` in dlpack is `int` data type.

EikanWang commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,90 @@
+  device_global_indexs.resize(num_gpus);
+}
+
+void initDeviceProperty(int device) {
```
```suggestion
void initDeviceProperty(DeviceIndex device) {
```

guangyey replied:
done.

EikanWang commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,90 @@
+  device_properties[device] = device_prop;
+}
+
+void initDeviceGlobalId(int device) {
```
```suggestion
void initDeviceGlobalIdx(DeviceIndex device) {
```

guangyey replied:
done.

EikanWang commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,90 @@
+    return raw_device == device;
+  };
+  auto it = std::find_if(devices.begin(), devices.end(), match_device);
+  TORCH_CHECK(it != devices.end(), "Cant't find the global id of XPU device.");
```
```suggestion
  TORCH_CHECK(it != devices.end(), "Cant't find the global index of XPU device.");
```

guangyey replied:
done.

EikanWang commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,90 @@
+  return getDeviceProperties(device);
+}
+
+DeviceProp* getDeviceProperties(int device) {
```
```suggestion
DeviceProp* getDeviceProperties(DeviceIndex device) {
```

guangyey replied:
done.

EikanWang commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,90 @@
+  return &device_properties[device];
+}
+
+int getGlobalIdxFromDevice(int device) {
```
```suggestion
DeviceIndex getGlobalIdxFromDevice(DeviceIndex device) {
```

guangyey replied:
done.

EikanWang commented on aten/src/ATen/xpu/detail/XPUHooks.h:
```diff
@@ -0,0 +1,21 @@
+  void initXPU() const override;
+  bool hasXPU() const override;
+  std::string showConfig() const override;
+  int getGlobalIdxFromDevice(const at::Device& device) const override;
```
```suggestion
  DeviceIndex getGlobalIdxFromDevice(const at::Device& device) const override;
```

guangyey replied:
done.

EikanWang commented on aten/src/ATen/xpu/XPUContext.h:
```diff
@@ -0,0 +1,18 @@
+
+TORCH_XPU_API DeviceProp* getCurrentDeviceProperties();
+
+TORCH_XPU_API DeviceProp* getDeviceProperties(int device);
+
+TORCH_XPU_API int getGlobalIdxFromDevice(int device);
```
The two APIs are global, right? Then, should we use INT to align with other backends? Or should we use unsigned or `DeviceIndex`?

guangyey replied:
Align with other backends. use `int` in public APIs.

guangyey committed:
[2/4] Intel GPU Runtime Upstreaming for Device

[ghstack-poisoned]

EikanWang commented on aten/src/ATen/detail/XPUHooksInterface.h:
```diff
@@ -9,11 +9,6 @@
 #include <functional>
 #include <memory>
 
-// We use forward declaration here instead of #include <ATen/dlpack.h> to avoid
-// leaking DLPack implementation detail to every project that includes `ATen/Context.h`, which in turn
-// would lead to a conflict when linked with another project using DLPack (for example TVM)
-struct DLDevice_;
-
```
These changes should not be part of this PR.

guangyey replied:
As we decouple `DLDevice_` from `XPUHooks`, here we remove it.

EikanWang commented on aten/src/ATen/detail/XPUHooksInterface.h:
```diff
@@ -44,23 +39,8 @@ struct TORCH_API XPUHooksInterface {
         XPU_HELP);
   }
 
-  virtual Device getATenDeviceFromDLPackDevice(
-      const DLDevice_& dl_device,
-      void* data) const {
-    TORCH_CHECK(
-        false,
-        "Cannot get XPU device without Intel Extension for Pytorch. ",
-        XPU_HELP);
-  }
-
-  virtual DLDevice_& getDLPackDeviceFromATenDevice(
-      DLDevice_& dl_device,
-      const Device& aten_device,
-      void* data) const {
-    TORCH_CHECK(
-        false,
-        "Cannot get XPU DL device without Intel Extension for Pytorch. ",
-        XPU_HELP);
```
Ditto

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"




[ghstack-poisoned]

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

gujinghui commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,93 @@
+  device_properties[device] = device_prop;
+}
+
+void initDeviceGlobalIdxs(int device) {
```
The plural of index should be indexes? initDeviceGlobalIdx is good enough.

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -101,8 +101,9 @@ static DLDevice getDLDevice(const Tensor& tensor, const int64_t& device_id) {
-          ctx, tensor.device(), tensor.data_ptr());
+      ctx.device_type = DLDeviceType::kDLOneAPI;
+      ctx.device_id =
+          at::detail::getXPUHooks().getGlobalIdxFromDevice(tensor.device());
```
`kDLOneAPI` is a device type from DLPack, XPU uses it here. To avoid leaking DLPack implementation detail to every project that includes `ATen/Context.h`, we decouple `DLDevice_` from `XPUHooks` and only use the index of `DLDevice_` as our input.

guangyey commented on aten/src/ATen/detail/XPUHooksInterface.h:
```diff
@@ -44,23 +39,8 @@ struct TORCH_API XPUHooksInterface {
-        XPU_HELP);
-  }
-
-  virtual DLDevice_& getDLPackDeviceFromATenDevice(
```
We replaced `getDLPackDeviceFromATenDevice` with `getGlobalIdxFromDevice`

guangyey commented on aten/src/ATen/detail/XPUHooksInterface.h:
```diff
@@ -44,23 +39,8 @@ struct TORCH_API XPUHooksInterface {
         XPU_HELP);
   }
 
-  virtual Device getATenDeviceFromDLPackDevice(
```
We replace `getATenDeviceFromDLPackDevice` with `getDeviceFromPtr`

guangyey commented on aten/src/ATen/dlpack.h:
```diff
@@ -94,10 +94,7 @@ typedef enum {
-// NB: This is the only difference from
-// https://github.com/dmlc/dlpack/blob/v0.7/include/dlpack/dlpack.h Required to
-// allow forward declaration of DLDevice.
-typedef struct DLDevice_ {
```
No need for forward declaration for `XPUHooksInterface` anymore, revert `DLDevice_` to its original state.

jgong5 commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -129,7 +130,7 @@ static Device getATenDevice(const DLDevice& ctx, void* data) {
 #endif
     case DLDeviceType::kDLOneAPI:
-      return at::detail::getXPUHooks().getATenDeviceFromDLPackDevice(ctx, data);
+      return at::detail::getXPUHooks().getDeviceFromPtr(data);
```
Can you explain why we cannot do `at::Device(DeviceType::XPU, ctx.device_id)` and has to be different from other devices here?

guangyey replied:
The main reason is that the SYCL specification does not explicitly state that the order of enumeration for SYCL devices is fixed. The external program may not necessarily correspond to the same order as device index in PyTorch. So we have to be cautious about this and find the target device via data pointer.
Please correct me if I am wrong.@gujinghui

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] [2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

huydhn commented:
@pytorch rebase

huydhn commented:
You can comment `@pytorchbot rebase` to rebase the PR to main.

gujinghui commented on aten/src/ATen/xpu/detail/XPUHooks.cpp:
```diff
@@ -0,0 +1,35 @@
+  return "XPU backend";
+}
+
+DeviceIndex XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
```
Should we use int as return type? The global index is not DeviceIndex defined by PyTorch, right?
Instead, it's defined by DLPack? We cannot guarantee it always aligns to PyTorch definition?

EikanWang replied:
@guangyey , as we discussed, we should return int and add some comments here, right?

guangyey replied:
Agreed. change to return int.

guangyey commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,93 @@
+std::vector<DeviceProp> device_properties;
+
+std::deque<c10::once_flag> device_global_idx_flags;
+std::vector<int> device_global_idxs;
```
`device_global_idxs` is used in DLPack, so use `int` as its data type.

malfet replied:
Would be good to be a bit more explicit, for example use `int32_t`

EikanWang commented:
> You can comment `@pytorchbot rebase` to rebase the PR to main.

Thanks. Yes, we will rebase the PR as soon as the first PR(#116019 ) being landed

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented:
@pytorchbot rebase

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

EikanWang commented:
@malfet , may I know if you have any comments on this ghstack?

chuanqi129 commented:
Hi @atalman @huydhn, seems the xpu test workflow has permission issue https://github.com/pytorch/pytorch/actions/runs/7502988971, could you please help to double check it?
```
Invalid workflow file: .github/workflows/xpu.yml#L29
The workflow is not valid. .github/workflows/xpu.yml (Line: 29, Col: 3): Error calling workflow 'pytorch/pytorch/.github/workflows/_xpu-test.yml@8a32701c9d734128e7f891615ff956d3fa1406c0'. The workflow is requesting 'id-token: write', but is only allowed 'id-token: none'.
```

guangyey commented:
@pytorchbot rebase

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey commented on aten/src/ATen/Context.h:
```diff
@@ -477,19 +479,6 @@ static inline void manual_seed(uint64_t seed) {
     }
   }
 
-  const auto xpu_num_gpus = detail::getXPUHooks().getNumGPUs();
```
These codes will be added with PR related to `Generator` as they affect `torch.manual_seed` in CI.

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

malfet commented:
Changed the topic, as "Not user facing" feels like a wrong topic for PRs that attempts to introduce new features, ins't it?

EikanWang commented:
> Changed the topic, as "Not user facing" feels like a wrong topic for PRs that attempts to introduce new features, ins't it?

Oh, yes. You mean users need to be aware of `torch.xpu`, right?

malfet commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,91 @@
+}
+
+void initDeviceProperty(DeviceIndex device) {
+  DeviceProp device_prop;
+  c10::xpu::get_device_properties(&device_prop, device);
+  device_properties[device] = device_prop;
```
Why not do that?
```suggestion
  c10::xpu::get_device_properties(&device_properties[device], device);
```

guangyey replied:
Thanks, done.

malfet commented on aten/src/ATen/xpu/XPUContext.cpp:
```diff
@@ -0,0 +1,91 @@
+      "device is out of range, device is ",
+      static_cast<int>(device),
+      ", total number of device is ",
+      static_cast<int>(num_gpus),
```
Why do you need to cast here?
```suggestion
      num_gpus,
```

guangyey replied:
`device` and `num_gpus` are `DeviceIndex` types, which is `int8_t`. `int8_t` is similar as `signed char` and `std::cout` can not correctly print its value. refer to https://stackoverflow.com/questions/7587782/why-is-stdcout-not-printing-the-correct-value-for-my-int8-t-number
According to the existing PyTorch [code](https://github.com/pytorch/pytorch/blob/1872834247084d76544c95e04fe2ec3370bac87d/c10/core/DeviceType.cpp#L58) in `c10/core/DeviceType.cpp`, I will change `static_cast<int>(num_gpus)` to `static_cast<int16_t>(num_gpus)`.

malfet reviewed (approved):
Looks OK to me, though it feels it mostly just copy-n-pastes code from `aten/src/ATen/cuda/` and renames `CUDA` to `XPU` 
If this is the case, would it be possible to use templates of sort to avoid code duplication?

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), as mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey committed:
Update on "[2/4] Intel GPU Runtime Upstreaming for Device"


# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), as mentioned in [[RFC] Intel GPU Runtime Upstreaming](https://github.com/pytorch/pytorch/issues/114842), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdxFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

guangyey commented:
@pytorchbot merge

guangyey commented:
@pytorchbot merge
